### ISSUE TITLE
Possible bug in random.laplace

### ISSUE THREAD

**Issue by yxwangcs** [numpy/numpy#13361]
On reviewing the code of `numpy/random/mtrand/distributions.c`, we noticed that in `rk_standard_exponential`, a trick is used to mitigate the fact that `rk_double` returns [0, 1).

https://github.com/numpy/numpy/blob/8f31f95910d27da75941bbbbde7ef1ecec9b0f74/numpy/random/mtrand/distributions.c#L110-L113

but why isn't the same trick used in rk_laplace?
https://github.com/numpy/numpy/blob/8f31f95910d27da75941bbbbde7ef1ecec9b0f74/numpy/random/mtrand/distributions.c#L664-L677

This may lead to problem if `U == 0` since `log(U + U)` will give `-inf`.

CC: @dingsquared

### Numpy/Python version information:
Current `master` (8f31f95910d27da75941bbbbde7ef1ecec9b0f74)

**Comment by EliasRojasC** [comment]
I would like to try this. Can I?

**Comment by RyanWangGit** [comment]
Sure, how do you plan to solve this? I think generally we want a `rk_double` that returns `(0, 1)` to solve this case, simply using `1.0 - U` would result in `2.0 - U - U == 0` if `U == 0`.

BTW, this issue also exists in 
https://github.com/numpy/numpy/blob/8f31f95910d27da75941bbbbde7ef1ecec9b0f74/numpy/random/mtrand/distributions.c#L687-L693
and maybe more distributions.

**Comment by bashtage** [comment]
I think a solution would be 
```
double rk_laplace(rk_state *state, double loc, double scale) 
 { 
     double U; 
  
     U = rk_double(state); 
     if (U == 0.0) 
     {
        return rk_laplace(state, loc, scale) 
     } 
     else if (U < 0.5) 
     { 
         U = loc + scale * log(U + U); 
     } else 
     { 
         U = loc - scale * log(2.0 - U - U); 
     } 
     return U; 
 } 
```

this is a stream change (since it becomes a rejection sampler) and so is difficult to implement in RandomState.  It should be fixed in @mattip 's randomgen branch, #13163.  Probably faster to switch the if-else if-else order. 

The same solution works for logistic.

**Comment by dingsquared** [comment]
I think we can do the following: break [0,1) into two non-overlapping
intervals [0,0.5) and [0.5, 1).
If U < 0.5, then 2U is in [0,1) so (1 - 2U) is in (0,1] and we use  -log( 1
- 2U) to get an exponential random variable.
Otherwise, 2U is in [1,2) so  (2 - 2U ) is in (0,1] so we use -log(2 - 2U).
The code will be something like

double rk_laplace(rk_state *state, double loc, double scale)
 {
     double U;

     U = rk_double(state);
     if (U < 0.5)
     {
         U = loc + scale * log(1.0 - U - U);
     } else
     {
         U = loc - scale * log(2.0 - U - U);
     }
     return U;
 }


Zeyu

**Comment by bashtage** [comment]
@dingsquared this could work in the new generator in randomgen but not for RandomState since it solution badly breaks the stream. The option of trapping NaNs might be OK since this is a bug and so a bad value.

**Comment by seberg** [comment]
@bashtage I am reading this as: too minor to risk breaking the stream for? So that the plan would be to fix it in 1.7 probably where we do not mind breaking the stream, but not in the 1.16.x minor release?

And if that is the decision, this issue is part of, and will be fixed by, gh-13163.

**Comment by bashtage** [comment]
@seberg RandomState as a compatibility guarantee and so a big change for a 1 in 10**53 event isn't worth it.  However, the rejection sampling will produce the same stream except when the bug occurs, in which case the nan would be replaced with the next value.

In the new branch, this fix which avoids recursion could be used since the stream isn't guaranteed.

**Comment by bashtage** [comment]
Also a bug in gumbel:

```c
  U = 1.0 - next_double(brng_state);
  return loc - scale * log(-log(U));
```

when U == 1.0 (or when the next_double is 0.0)

**Comment by bashtage** [comment]
Closed by #13416
